### PR TITLE
feat(perp): use peg multiplier to compute prices

### PR DESCRIPTION
**Pull request by matthiasmatt**
# Description

Use peg and bias to compute the price of a pool.

# Purpose

New logic that can be repegged

### CODE REVIEW

File: proto/perp/amm/v1/state.proto [hunk 1]
```diff
@@ -145,7 +145,19 @@ message ReserveSnapshot {
   ];
 
   // milliseconds since unix epoch
-  int64 timestamp_ms = 3;
```
Comment (jgimeno) [inline proto/perp/amm/v1/state.proto:148]: Do you know why there is this change in number? In theory this is not recommended.

File: x/perp/amm/abci_test.go [hunk 2]
```diff
@@ -31,21 +31,21 @@ func TestSnapshotUpdates(t *testing.T) {
...
-		*types.DefaultMarketConfig().
-			WithTradeLimitRatio(sdk.OneDec()).
-			WithFluctuationLimitRatio(sdk.OneDec()),
-		sdk.ZeroDec(),
-		sdk.OneDec(),
+		/* ctx */ ctx,
+		/* pair */ asset.Registry.Pair(denoms.BTC, denoms.NUSD),
+		/* quoteReserve */ sdk.NewDec(1_000),
+		/* baseReserve */ sdk.NewDec(1_000),
+		/* config */ *types.DefaultMarketConfig().WithTradeLimitRatio(sdk.OneDec()).WithFluctuationLimitRatio(sdk.OneDec()),
```
Comment (jgimeno) [inline x/perp/amm/abci_test.go:38]: I would set defaultmarketconfig as type, not pointer, to avoid this casting.

**Review by jgimeno — APPROVED** [review]

File: x/perp/amm/abci_test.go (hunk 2 above)
Comment (Unique-Divine) [reply x/perp/amm/abci_test.go:38]: Golang doesn't have the default args feature, you have have to use pointer if you want to see default values inside a constructor

File: x/perp/amm/keeper/bias_test.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func TestBiasChangeOnMarket(t *testing.T) {
 			).
 			Then(
 				ammassertion.MarketShouldBeEqual(pairBtcUsdc,
-					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("9999.999900000001000000")), // Bias equal to PositionSize
+					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("-9999.999900000001000000")), // Bias equal to PositionSize
 				),
```
Comment (Unique-Divine) [inline x/perp/amm/keeper/bias_test.go:59]: These test cases are opposite. The trader opened long, so the bias should be positive. Semantically, we'd be saying the market has a long (positive) bias.

File: x/perp/amm/keeper/bias_test.go [hunk 4]
```diff
@@ -75,7 +75,7 @@ func TestBiasChangeOnMarket(t *testing.T) {
 			).
 			Then(
 				ammassertion.MarketShouldBeEqual(pairBtcUsdc,
-					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("19999.999600000008000000")), // Bias equal to PositionSize
```
Comment (Unique-Divine) [inline x/perp/amm/keeper/bias_test.go:78]: same (opposite)

File: x/perp/amm/keeper/bias_test.go [hunk 5]
```diff
@@ -92,7 +92,7 @@ func TestBiasChangeOnMarket(t *testing.T) {
 			).
 			Then(
 				ammassertion.MarketShouldBeEqual(pairBtcUsdc,
-					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("-10000.000100000001000000")), // Bias equal to PositionSize
+					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("10000.000100000001000000")), // Bias equal to PositionSize
 				),
 				PositionShouldBeEqual(alice, pairBtcUsdc, Position_PositionSizeShouldBeEqualTo(sdk.MustNewDecFromStr("-10000.000100000001000000"))),
```
Comment (Unique-Divine) [inline x/perp/amm/keeper/bias_test.go:97]: same (opposite)

File: x/perp/amm/keeper/bias_test.go [hunk 6]
```diff
@@ -112,7 +112,7 @@ func TestBiasChangeOnMarket(t *testing.T) {
 			).
 			Then(
 				ammassertion.MarketShouldBeEqual(pairBtcUsdc,
-					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("-20000.000400000008000000")), // Bias equal to PositionSize
+					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("20000.000400000008000000")), // Bias equal to PositionSize
 				),
```
Comment (Unique-Divine) [inline x/perp/amm/keeper/bias_test.go:116]: same (opposite)

File: x/perp/amm/keeper/bias_test.go [hunk 7]
```diff
@@ -150,7 +150,7 @@ func TestBiasChangeOnMarket(t *testing.T) {
 			).
 			Then(
 				ammassertion.MarketShouldBeEqual(pairBtcUsdc,
-					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("8999.999919000000729000")), // Bias equal to PositionSize
+					ammassertion.Market_BiasShouldBeEqualTo(sdk.MustNewDecFromStr("-8999.999919000000729000")), // Bias equal to PositionSize
 				),
```
Comment (Unique-Divine) [inline x/perp/amm/keeper/bias_test.go:154]: same (opposite)

File: x/perp/amm/keeper/prices.go [hunk 8]
```diff
@@ -59,37 +59,12 @@ func (k Keeper) GetBaseAssetPrice(
...
-	pool, err := k.Pools.Get(ctx, pair)
+	quoteAmount, err = amm.GetQuoteAmountByBaseAmount(baseAssetAmount.MulInt64(dir.ToMultiplier()))
 	if err != nil {
-		return sdk.ZeroDec(), err
+		return
 	}
-
-	dirMult := dir.ToMultiplier()
-	return pool.GetBaseAmountByQuoteAmount(quoteAmountAbs.MulInt64(dirMult))
+	quoteAmount = quoteAmount.Mul(amm.PegMultiplier)
```
Comment (Unique-Divine) [inline x/perp/amm/keeper/prices.go:66]: It looks like we were supposed to multiply by peg, then divide base. Is the input arg in assets or reserves?

**Review by Unique-Divine — CHANGES_REQUESTED** [review]
Review still WIP, coming back shortly

File: proto/perp/amm/v1/state.proto (hunk 1 above)
Comment (k-yang) [reply proto/perp/amm/v1/state.proto:148]: 3 and 5 are missing field numbers.

File: x/perp/amm/abci_test.go (hunk 2 above)
Comment (k-yang) [reply x/perp/amm/abci_test.go:38]: It should be possible to pass a struct instead of a pointer to a struct.

**Review by k-yang — APPROVED** [review]

**Pull request merged**